Make it illegal to directly construct _TensorBase

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56150 Make it illegal to directly construct _TensorBase**
* #56147 Give Tensor a trivial (for now) metaclass _TensorMeta

See #56017 for full context; the short story is that by making
it illegal to directly construct _TensorBase, we need only
write a *single* tp_dealloc function which will work universally
for all _TensorBase subclasses, rather than having to write two
versions, one for _TensorBase itself, and others for Python subclasses
of _TensorBase.  This means simpler code.

The subtlety here is that we only install our custom `tp_new` for direct subclasses of TensorBase.  This is important, because overriding the `tp_new` also overrides any user defined constructor.  Fortunately class Tensor(_TensorBase) has no nontrivial constructors and doesn't mind, but other subclasses like Parameter definitely mind!

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28028746](https://our.internmc.facebook.com/intern/diff/D28028746)

ezyang committed:
Make it illegal to directly construct _TensorBase

See #56017 for full context; the short story is that by making
it illegal to directly construct _TensorBase, we need only
write a *single* tp_dealloc function which will work universally
for all _TensorBase subclasses, rather than having to write two
versions, one for _TensorBase itself, and others for Python subclasses
of _TensorBase.  This means simpler code.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make it illegal to directly construct _TensorBase"


See #56017 for full context; the short story is that by making
it illegal to directly construct _TensorBase, we need only
write a *single* tp_dealloc function which will work universally
for all _TensorBase subclasses, rather than having to write two
versions, one for _TensorBase itself, and others for Python subclasses
of _TensorBase.  This means simpler code.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make it illegal to directly construct _TensorBase"


See #56017 for full context; the short story is that by making
it illegal to directly construct _TensorBase, we need only
write a *single* tp_dealloc function which will work universally
for all _TensorBase subclasses, rather than having to write two
versions, one for _TensorBase itself, and others for Python subclasses
of _TensorBase.  This means simpler code.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -909,9 +903,22 @@ PyTypeObject THPVariableType = {
+  if (PyType_Type.tp_init(cls, args, kwargs) < 0) {
+    return -1;
+  }
+  if (((PyTypeObject*)cls)->tp_base == &THPVariableType) {
```
This means that the class being created is a subclass of _TensorBase? If so, it doesn't match the description in the PR where you seem to say that only the _TensorBase itself should have its tp_new modified?

ezyang replied:
I'm not sure what PR description you're referring to. I did get this wrong the first time I wrote the PR, the operant clarification is:

> The subtlety here is that we only install our custom tp_new for direct subclasses of TensorBase. This is important, because overriding the tp_new also overrides any user defined constructor. Fortunately class Tensor(_TensorBase) has no nontrivial constructors and doesn't mind, but other subclasses like Parameter definitely mind!

The logic here is testing for a direct subclass of TensorBase.

albanD replied:
Ho right, THPVariable is not Tensor, but _TensorBase. So this match the Tensor class!
Sounds good!

albanD reviewed (approved):
LGTM!

ezyang committed:
Update on "Make it illegal to directly construct _TensorBase"


See #56017 for full context; the short story is that by making
it illegal to directly construct _TensorBase, we need only
write a *single* tp_dealloc function which will work universally
for all _TensorBase subclasses, rather than having to write two
versions, one for _TensorBase itself, and others for Python subclasses
of _TensorBase.  This means simpler code.

The subtlety here is that we only install our custom `tp_new` for direct subclasses of TensorBase.  This is important, because overriding the `tp_new` also overrides any user defined constructor.  Fortunately class Tensor(_TensorBase) has no nontrivial constructors and doesn't mind, but other subclasses like Parameter definitely mind!

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28028746](https://our.internmc.facebook.com/intern/diff/D28028746)

[ghstack-poisoned]
